### PR TITLE
Avoid potential for infinite blocking in IT

### PR DESCRIPTION
**Pull request by SamBarker**
### Type of change

- Bugfix

### Description
Apply a timeout to `Future.get` calls in the MultiTenantIT and wrap the Kafka clients in a wrapper which intercepts the call to close and calls it with a much more appropriate timeout.

### Additional Context
The multiTenantIT can block for what amounts to forever. I found this the hard way debugging @racheljpg's transactions IT. What makes things rather insidious is that even timing out the producer future doesn't let the test complete as by the Kafka Clients implement close as `close(Duration.ofMillis(Long.MAX_VALUE))` which means the try-with-resources statements we use for managing clients block indefinitely (and magically as its done implicitly)

If there is consensus that this is the right approach then we should wrap clients in other tests using the same fix.

### CODE REVIEW

**Comment by k-wall** [comment]
> What makes things rather insidious is that even timing out the producer future doesn't let the test complete as by the Kafka Clients implement close as close(Duration.ofMillis(Long.MAX_VALUE)) which means the try-with-resources statements we use for managing clients block indefinitely (and magically as its done implicitly)

Ha!  Yes, I was discovered the same thing yesterday working on #254 . The defect was causing a produce to a follower, which goes into an infinite loop and blocks #close() forever.  I also discovered that canceling the produce future does nothing.

I'm liking your idea. Adding @showuon as a reviewer in case he has thoughts from a Kafka Client API angle.

File: kroxylicious/pom.xml [hunk 1]
```diff
@@ -332,10 +332,31 @@
         </profile>
         <profile>
             <id>debug</id>
+            <dependencyManagement>
+                <dependencies>
+                    <dependency>
+                        <groupId>org.apache.logging.log4j</groupId>
+                        <artifactId>log4j-bom</artifactId>
+                        <version>2.20.0</version>
```
Comment (k-wall) [inline kroxylicious/pom.xml:340]: Please extract property.

**Review by k-wall — APPROVED** [review]

File: integrationtests/src/test/java/io/kroxylicious/proxy/CloseableAdmin.java [hunk 2]
```diff
@@ -0,0 +1,644 @@
...
+
+public record CloseableAdmin(Admin instance) implements Admin, AutoCloseable {
+
+    public static Admin wrap(Admin instance) {
+        return new CloseableAdmin(instance);
+    }
+
+    @Override
+    public void close() {
+        instance.close(Duration.ofSeconds(5L));
```
Comment (k-wall) [inline integrationtests/src/test/java/io/kroxylicious/proxy/CloseableAdmin.java:137]: Wondering whether these decorator belong in an common library.  Maybe in the junit ext?

**Comment by showuon** [comment]
> Ha! Yes, I was discovered the same thing yesterday working on #254 . The defect was causing a produce to a follower, which goes into an infinite loop and blocks #close() forever. I also discovered that canceling the produce future does nothing.
> 
> I'm liking your idea. Adding @showuon as a reviewer in case he has thoughts from a Kafka Client API angle

Sam shared the thought with me today and I also agree with it. Sam will raise a JIRA issue next week.

File: kroxylicious/pom.xml (hunk 1 above)
Comment (SamBarker) [reply kroxylicious/pom.xml:340]: Out of interest why? We already have a place defining all the versions of the log4j dependencies (here), we don't really want people to add additional ones with manually managed versions.

Looking again i didn't use import scope here so need to fix it anyway

Comment (SamBarker) [reply kroxylicious/pom.xml:340]: Oh no I did just the diff view on my phone being poor

File: integrationtests/src/test/java/io/kroxylicious/proxy/CloseableAdmin.java (hunk 2 above)
Comment (SamBarker) [reply integrationtests/src/test/java/io/kroxylicious/proxy/CloseableAdmin.java:137]: That's an interesting thought, that is probably a better home.

**Comment by racheljpg** [comment]
Hey @SamBarker 
I'm sorry, I'm not really understanding from your comment/change why exactly the `close(Duration.ofMillis(Long.MAX_VALUE))` line causes things to be blocked infinitely. Where's the max value coming from? Why is it not timing out after that? I'm a bit lost :)

**Comment by SamBarker** [comment]
Hey @racheljpg 
Lots of reasonable questions, I'll answer them in turn hopefully they shed some light onto what's going on.

> I'm sorry, I'm not really understanding from your comment/change why exactly the `close(Duration.ofMillis(Long.MAX_VALUE))` line causes things to be blocked infinitely.

So a try-with-resources statement requires it to be passed something that implements `Closeable` or `AutoCloseable`. Both those interfaces specify a method `close()`. Handily for us the Kafka clients do implement `AuotCloseable` so we can use them as a resources in try-with-resources. Yay! The clients implement `close()` by calling the overloaded method signature `close(Duration waitFor)`, which waits for up to `waitFor` to allow pending operations to complete. 

When everything works normally the the JVM will call `close()` as the try block completes and the client will disconnect in response as it has nothing outstanding. 

When things go wrong (as in the current state of the Transactions test you are looking at) the Kafka client waits for the pending operations to complete, however due to the nature of the failure they will never complete successfully. So the Kafka client is going to wait until `Duration.ofMillis(Long.MAX_VALUE)` for the pending operation to complete successfully. 

> Where's the max value coming from?

The max is hard coded into the [KafkaProducer](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java#L1277-L1280). I'm planning to raise a PR with Apache Kafka to allow it to be configured via a system property. 

>Why is it not timing out after that?

Converting `Long.MAX_VALUE=2^63-1` (why `-1`? Because in Java we are dealing with signed values) into a duration is roughly 292471208 years (I don't know about anyone else but I'm not waiting that long for a test to timeout).

**Review by showuon — COMMENTED** [review]
Using the wrap looks good to me

File: .idea/misc.xml [hunk 3]
```diff
@@ -11,5 +11,5 @@
   <component name="PDMPlugin">
     <option name="skipTestSources" value="false" />
   </component>
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_17" default="true" project-jdk-name="temurin-17" project-jdk-type="JavaSDK" />
+  <component name="ProjectRootManager" version="2" languageLevel="JDK_19" default="true" project-jdk-name="temurin-19" project-jdk-type="JavaSDK" />
```
[the remark this answers is not in the document]
Comment (SamBarker) [reply .idea/misc.xml:14]: Yeah 😕

To configure IntelliJ to build any module with Java 19 the project SDK needs to be jdk 19. IntelliJ then correctly figures out the language level within each module

**Comment by k-wall** [comment]
Now we've got the release done, presumably we can brush up this PR to adopt the new stuff

**Pull request merged**